### PR TITLE
Support BlurStyle values in BoxShadow (#52999)

### PR DESCRIPTION
**Pull request by deepak-penaganti**
*`BoxShadow` does not support all the `BlurStyle`s*

*Adds support for shadow types as mentioned in issue #52999*

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by deepak-penaganti** [comment]
@googlebot I signed it!

**Comment by HansMuller** [comment]
This PR will need tests.

@GaryQian - not sure who the right reviewer would be.

**Comment by Piinks** [comment]
(triage) @deepak-penaganti would you like to continue with this change and add tests?

**Comment by deepak-penaganti** [comment]
> (triage) @deepak-penaganti would you like to continue with this change and add tests?

Will add tests

**Comment by deepak-penaganti** [comment]
@Piinks Added tests for the change

File: packages/flutter/lib/src/painting/box_shadow.dart [hunk 1]
```diff
@@ -36,11 +36,15 @@ class BoxShadow extends ui.Shadow {
     Offset offset = Offset.zero,
     double blurRadius = 0.0,
     this.spreadRadius = 0.0,
+    this.blurStyle = BlurStyle.normal,
   }) : super(color: color, offset: offset, blurRadius: blurRadius);
 
   /// The amount the box should be inflated prior to applying the blur.
   final double spreadRadius;
 
+  /// The style(inner/normal/outer/solid) of the blur.
```
Comment (GaryQian) [inline packages/flutter/lib/src/painting/box_shadow.dart:45]: Let's keep the formatting/doc style consistent with the rest of Flutter. How about:

```
/// The [BlurStyle] to use for this shadow.
///
/// Defaults to [BlurStyle.normal].
```

Comment (deepak-penaganti) [reply packages/flutter/lib/src/painting/box_shadow.dart:45]: Sure thing @GaryQian

**Comment by deepak-penaganti** [comment]
@iapicca @TahaTesser @darshankawar Please do get a look at this PR and assign [this](https://github.com/flutter/flutter/issues/52999) related issue

**Comment by GaryQian** [comment]
Ok, so the existing code all looks great, but since much of the rendering of the shadow happens in the engine, we should ensure all of the mechanisms that pass the style down are all functional and tested.

A great way to do this is a golden which can ensure the final drawn result has the correct shadow and visually looks correct.

Try something like:

```
testWidgets('BoxShadow BoxStyle.solid', (WidgetTester tester) async {
    final Key key = UniqueKey();
    await tester.pumpWidget(
      Center(
        child: RepaintBoundary(
          key: key,
          child: Container(
            color: Colors.white,
            width: 50,
            height: 50,
            child: Center(
              child: Container(
                decoration: BoxDecoration(
                  boxShadow: <BoxShadow>[BoxShadow(blurRadius: 3.0, blurStyle: BlurStyle.inner)],
                ),
                width: 10,
                height: 10,
              ),
            ),
          ),
        ),
      ),
    );

    await expectLater(
      find.byKey(key),
      matchesGoldenFile('boxShadow.boxStyle.inner.0.0.png'),
    );
  });
```

We can then use SkiaGold in the presubmit checks to commit the golden png images.

**Comment by skia-gold** [comment]
Gold has detected about 1 new digest(s) on patchset 9.
View them at https://flutter-gold.skia.org/cl/github/84733

**Comment by deepak-penaganti** [comment]
@GaryQian Not sure why **_customer_testing-linux_** keeps failing 🤔

**Comment by goderbauer** [comment]
> @GaryQian Not sure why customer_testing-linux keeps failing 🤔

Looks like a flake. I've restarted the test. If it continues to fail, you may have to rebase this branch on top of the latest master.

File: packages/flutter/test/painting/box_painter_test.dart [hunk 2]
```diff
@@ -107,7 +109,70 @@ void main() {
...
+    expect(shadowList[1].blurStyle, equals(BlurStyle.normal));
+  });
+
   test('BoxShadow toString test', () {
-    expect(const BoxShadow(blurRadius: 4.0).toString(), equals('BoxShadow(Color(0xff000000), Offset(0.0, 0.0), 4.0, 0.0)'));
+    expect(const BoxShadow(blurRadius: 4.0).toString(), equals('BoxShadow(Color(0xff000000), Offset(0.0, 0.0), 4.0, 0.0), BlurStyle.normal'));
+    expect(const BoxShadow(blurRadius: 4.0, blurStyle: BlurStyle.solid).toString(), equals('BoxShadow(Color(0xff000000), Offset(0.0, 0.0), 4.0, 0.0), BlurStyle.solid'));
+  });
+
+  testWidgets('BoxShadow BoxStyle.solid', (WidgetTester tester) async {
```
Comment (GaryQian) [inline packages/flutter/test/painting/box_painter_test.dart:149]: One last thing, while you are at it, it should be easy to copypaste this test a few times to cover all of the different BlurStyles!

**Comment by GaryQian** [comment]
Customer_testing-linux still seems to be failing, why don't you try rebasing? I don't think this change should cause any failures in that test shard.

`git pull upstream master --rebase`

**Comment by skia-gold** [comment]
Gold has detected about 1 new digest(s) on patchset 169.
View them at https://flutter-gold.skia.org/cl/github/84733

File: packages/flutter/test/painting/box_painter_test.dart (hunk 2 above)
Comment (deepak-penaganti) [reply packages/flutter/test/painting/box_painter_test.dart:149]: Done 😃

**Comment by skia-gold** [comment]
Gold has detected about 2 new digest(s) on patchset 170.
View them at https://flutter-gold.skia.org/cl/github/84733

**Comment by christopherfujino** [comment]
The diff on this PR looks wrong (it's so big I can't load it in my browser), likely due to either an incorrect branch point (not from master), or an incorrect merge (not from master). This PR cannot be landed as is.

**Review by christopherfujino — CHANGES_REQUESTED** [review]
Please fix the diff on this PR (it is too large)

**Comment by deepak-penaganti** [comment]
@christopherfujino I hear you, not sure how merge went wrong. Please do check this [PR#86504](https://github.com/flutter/flutter/pull/86504)

**Comment by deepak-penaganti** [comment]
We can close this in favour of #86504

**Comment by christopherfujino** [comment]
closing in favor of https://github.com/flutter/flutter/pull/86504

**Pull request closed without merging**